metagraph/src/graph/representation/masked_graph.cpp

```diff
@@ -12,26 +12,34 @@ MaskedDeBruijnGraph
 ::MaskedDeBruijnGraph(std::shared_ptr<const DeBruijnGraph> graph,
                       std::unique_ptr<bitmap>&& kmers_in_graph,
                       bool only_valid_nodes_in_mask,
-                      bool canonical)
+                      Mode mode)
       : graph_(graph),
         kmers_in_graph_(std::move(kmers_in_graph)),
         only_valid_nodes_in_mask_(only_valid_nodes_in_mask),
-        is_canonical_(canonical) {
+        mode_(mode) {
     assert(kmers_in_graph_.get());
     assert(kmers_in_graph_->size() == graph->max_index() + 1);
-    assert(!is_canonical_ || graph_->is_canonical_mode());
+
+    if (graph_->get_mode() == DeBruijnGraph::PRIMARY
+            && mode_ != DeBruijnGraph::PRIMARY) {
+        throw std::runtime_error("Any subgraph of a primary graph is primary");
```

Comment:
should we make this a warning instead of throwing? we can essentially just set the value ourselves and warn the user that we changed it

Reply:
By automatically correcting things that don't make sense, you allow writing code that doesn't make sense.

Just imagine, you would allow writing `MaskedGraph graph(..., BASIC)`, where `graph.get_mode() == BASIC` will return `false` in the next line.

Reply:
Yes, true. In that case, how about we change the message here to explicitly say that `mode` must be set to `PRIMARY` when constructed from a primary graph?

Reply:
done